rust/ql/lib/codeql/rust/controlflow/internal/SuccessorType.qll

```diff
@@ -0,0 +1,46 @@
+private import codeql.util.Boolean
+
+cached
+newtype TSuccessorType =
+  TSuccessorSuccessor() or
+  TBooleanSuccessor(Boolean b) or
+  TReturnSuccessor()
+
+/** The type of a control flow successor. */
+class SuccessorType extends TSuccessorType {
```

Comment:
Since this class is exposed, I would do
```ql
private abstract class SuccessorTypeImpl extends TSuccessorType {
  /** Gets a textual representation of successor type. */
  abstract string toString();
}

final class SuccessorType = SuccessorTypeImpl;

// ...

private abstract class ConditionalSuccessorImpl extends SuccessorType {
 // ...
}

final class ConditionalSuccessor = ConditionalSucessorImpl;

// etc (make all public classes final)
```